src/NLog/Internal/ProcessIDHelper.cs

```diff
@@ -31,55 +31,60 @@
 // THE POSSIBILITY OF SUCH DAMAGE.
 // 
 
-#if !SILVERLIGHT && !__IOS__ && !NETSTANDARD1_3
+#if !SILVERLIGHT && !NETSTANDARD1_3
 
 namespace NLog.Internal
 {
-    using NLog.Config;
+    using System.IO;
 
     /// <summary>
     /// Returns details about current process and thread in a portable manner.
     /// </summary>
-    internal abstract class ThreadIDHelper
+    internal abstract class ProcessIDHelper
     {
-        /// <summary>
-        /// Initializes static members of the ThreadIDHelper class.
-        /// </summary>
-        static ThreadIDHelper()
-        {
-#if !SILVERLIGHT && !__IOS__ && !__ANDROID__ && !NETSTANDARD
-            if (PlatformDetector.IsWin32)
-            {
-                Instance = new Win32ThreadIDHelper();
-            }
-            else
-#endif
-            {
-                Instance = new PortableThreadIDHelper();
-            }
-        }
+        private const string UnknownProcessName = "<unknown>";
+
+        private static ProcessIDHelper _threadIDHelper;
+        private string _currentProcessBaseName;
 
         /// <summary>
         /// Gets the singleton instance of PortableThreadIDHelper or
         /// Win32ThreadIDHelper depending on runtime environment.
         /// </summary>
         /// <value>The instance.</value>
-        public static ThreadIDHelper Instance { get; private set; }
+        public static ProcessIDHelper Instance => _threadIDHelper ?? (_threadIDHelper = CreateInstance());
 
         /// <summary>
         /// Gets current process ID.
         /// </summary>
         public abstract int CurrentProcessID { get; }
 
         /// <summary>
-        /// Gets current process name.
+        /// Gets current process absolute file path.
         /// </summary>
-        public abstract string CurrentProcessName { get; }
+        public abstract string CurrentProcessFilePath { get; }
 
         /// <summary>
         /// Gets current process name (excluding filename extension, if any).
         /// </summary>
-        public abstract string CurrentProcessBaseName { get; }
+        public string CurrentProcessBaseName => _currentProcessBaseName ?? (_currentProcessBaseName = string.IsNullOrEmpty(CurrentProcessFilePath) ? UnknownProcessName : Path.GetFileNameWithoutExtension(CurrentProcessFilePath));
+
+        /// <summary>
+        /// Initializes the ThreadIDHelper class.
+        /// </summary>
+        private static ProcessIDHelper CreateInstance()
```

Comment:
👍 much better. Please rename it to just "Create" (it's a factory method) - and what other to create ;)

Reply:
Think the whole instance thing should be removed. But not sure about the reason for the Win32-native-call. Since the values are cached, then I don't understand remark about Win32 is "without the overhead".